### PR TITLE
 infer lang and version

### PR DESCRIPTION
**Pull request by DanielMcSheehy**
https://app.clubhouse.io/effx/story/1182/extract-language-versioning

adds a `langugage:version` (and inferred tag annotation) for each service, if it can be detected
also adds a `language:<language>` tag as well as a `effx.io/inferred-tags` annotation. 

both of these tags will not overwrite customer data, if a custom already added those tags/fields. 

# currently supports 
- golang (go.mod)
- javascript (package.json)- writes the tag as `"node":"13.4"` (removing all inequalities) 
- php (composer.json)
- java [maven] (pom.xml) - there's three different things we can match on

# how to test
- I added test files for all language version packages

### CODE REVIEW

File: data/version.go [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+	} `json:"engines"`
+}
+
+type result struct {
+	lang    string
+	version string
+}
+
+// regex to find go version from go.mod file
+const GoVersionRegex = `go \d+(\.\d+)+`
```
Comment (weeyum) [inline data/version.go:24]: can we handle go and node in separate packages? kinda like how ventress has seperate handlers for each integration

Comment (weeyum) [reply data/version.go:24]: mainly worried about how big this file will get in the future

Comment (DanielMcSheehy) [reply data/version.go:24]: oh yeah, probably a good idea.

File: data/data.go [hunk 2]
```diff
@@ -40,12 +41,27 @@ func (y EffxYaml) newConfig() (*effx_api.ConfigurationFile, error) {
...
 		"effx.io/source":    "effx-cli",
 		"effx.io/file-path": y.FilePath,
 	})
 
+	if metadata != nil {
+		config.SetAnnotations(map[string]string{
+			"effx.io/cli-inferred-tags": metadata.lang,
+		})
+		config.SetTags(map[string]string{
+			metadata.lang: metadata.version,
```
Comment (weeyum) [inline data/data.go:61]: will this also set `language: <lang>` somewhere? we should be doing both, everywhere when possible

Comment (weeyum) [reply data/data.go:61]: can we make sure we are not overwriting someones tags/annotations? does that already happen?

Comment (DanielMcSheehy) [reply data/data.go]: good idea, setting langugage as a tag.

also I added something to ensure we are only appending to tags annotations, we won't wipe anything

File: data/data.go [hunk 3]
```diff
@@ -29,6 +31,23 @@ type EffxYaml struct {
...
+	if m != nil {
+		if config.Annotations == nil {
+			config.Annotations = &map[string]string{}
+		}
+		if config.Tags == nil {
+			config.Tags = &map[string]string{}
+		}
+
+		(*config.Annotations)["effx.io/inferred-tags"] = fmt.Sprintf("language,%s", m.Language)
+		(*config.Tags)["language"] = m.Language
```
Comment (weeyum) [inline data/data.go:44]: I'm concerned if the user has already defined a language field. we should take theirs instead. can we add some logic to only add the annotation and tags if they are undefined?

Comment (DanielMcSheehy) [reply data/data.go]: sure, I added a function that lets us safely write to those, without wiping anything. I tested it as well.

**Review by euriostigue — APPROVED** [review]

**Review by weeyum — APPROVED** [review]

**Pull request merged**